### ISSUE TITLE
lxc containers and proxmox cluster (+fix)

### ISSUE THREAD

**Issue by liedekef** [glpi-project/glpi-agent#463]
### Bug reporting acknowledgment

Yes, I read it

### Professional support

None

### Describe the bug

LXC containers managed via proxmox cluster and migrated away still leave a trace in /var/lib/lxc, causing lxc-ls to report containers no longer active/managed on that node.
The code tries to take proxmox into account (Task/Inventory/Virtualization/Lxc.pm), but the command for lxc-containers on proxmox should be "pct list" (and not "lxc-ls").

### To reproduce

1. create a lxc container
2. migrate to another proxmox node
3. lxc-ls on the first still reports it (in state shutdown)
4. the proxmox command "pct list" does not report it

### Expected behavior

For proxmox, the lxc-containers managed via proxmox should be taken.
Code change is quite simple: 

```
--- Lxc.pm.orig	2023-08-16 15:20:57.407678511 +0200
+++ Lxc.pm	2023-08-16 15:23:15.898465389 +0200
@@ -14,7 +14,9 @@
 use GLPI::Agent::Tools::Virtualization;
 
 sub isEnabled {
-    return canRun('lxc-ls');
+    return
+	canRun('lxc-ls') ||
+        canRun('pct');
 }
 
 sub doInventory {
@@ -124,10 +126,19 @@
 sub  _getVirtualMachines {
     my (%params) = @_;
 
-    my @lines = getAllLines(
-        command => 'lxc-ls -1',
-        %params
-    );
+    my @lines;
+
+    if (canRun('pct')) {
+	    @lines = getAllLines(
+			    command => 'pct list|tail -n +2|cut -d" " -f1',
+			    %params
+			    );
+    } else {
+	    @lines = getAllLines(
+			    command => 'lxc-ls -1',
+			    %params
+			    );
+    }
     return unless @lines;
 
     my $version = getFirstMatch(

```

If you agree on this, I can create a pull request for this.

### Operating system

Linux

### GLPI Agent version

Nightly build (git version in additional context below)

### GLPI version

10.0.6

### Additional context

version: 1.5-gitfd04a0fc

**Comment by g-bougard** [comment]
Hi @liedekef 

thank you for your detailed report.

Actually running few piped commands is not the way of doing. You need to just run the `pct list` command and use perl code to parse the output.

If you have some output samples, I can implement it if you want including your samples in unit tests.

**Comment by erik-de-neve** [comment]
Here 2 examples:

```
# pct list
VMID       Status     Lock         Name
106        running                 srv-b-worker-1
119        stopped    create       CT119
127        running                 srv-b-manager

# pct list
VMID       Status     Lock         Name
109        running                 srv-b-worker-2 
```

Kr,
Erik

**Comment by g-bougard** [comment]
Hi @erik-de-neve 

thank you for the outputs.

can you try the following patch ?
```diff
diff --git a/lib/GLPI/Agent/Task/Inventory/Virtualization/Lxc.pm b/lib/GLPI/Agent/Task/Inventory/Virtualization/Lxc.pm
index 22b658752..1447dbb67 100644
--- a/lib/GLPI/Agent/Task/Inventory/Virtualization/Lxc.pm
+++ b/lib/GLPI/Agent/Task/Inventory/Virtualization/Lxc.pm
@@ -14,7 +14,7 @@ use GLPI::Agent::Tools::Network;
 use GLPI::Agent::Tools::Virtualization;
 
 sub isEnabled {
-    return canRun('lxc-ls');
+    return canRun('lxc-ls') || canRun('pct');
 }
 
 sub doInventory {
@@ -125,8 +125,12 @@ sub  _getVirtualMachine {
 sub  _getVirtualMachines {
     my (%params) = @_;
 
+    # Check if we require to list containers using proxmox pct command
+    my $use_pct = canRun('pct');
+    my $pct_name = 0;
+
     my @lines = getAllLines(
-        command => 'lxc-ls -1',
+        command => $use_pct ? 'pct list' : 'lxc-ls -1',
         %params
     );
     return unless @lines;
@@ -148,6 +152,18 @@ sub  _getVirtualMachines {
     my @machines;
 
     foreach my $name (@lines) {
+        # Support pct when running with proxmox
+        if ($use_pct) {
+            if ($name =~ /^(VMID\s.*\s)Name$/) {
+                $pct_name = length($1);
+                next;
+            } elsif ($pct_name) {
+                $name = substr($name, $pct_name);
+            } else {
+                next;
+            }
+        }
+
         # lxc-ls -1 shows one entry by line
         $name =~ s/\s+$//;         # trim trailing whitespace
         next unless length($name); # skip if empty as name can contain space
```

**Comment by g-bougard** [comment]
Hi @liedekef 

did you have time to try my patch ?

**Comment by liedekef** [comment]
I'm leaving the testing to @erik-de-neve (currently unable to do any testing)

**Comment by g-bougard** [comment]
Hi @erik-de-neve 
were you able to test my patch ?

**Comment by yoshi314** [comment]
i'll test this on proxmox install i have access to.

**Comment by g-bougard** [comment]
> i'll test this on proxmox install i have access to.

Any feedback @yoshi314 ?

@liedekef @erik-de-neve did you make the test ?

I really want to know if I can include this patch as a fix before the next release which will arise in a week or 2.

**Comment by liedekef** [comment]
I'm no longer able to test this (left the company) and Erik seems to be on holiday this week.

**Comment by yoshi314** [comment]
It works, but i had to change a few things. ( I applied it to fusioninventory-agent, since i'm still using that one).

The issue with that proxmox uses numeric ids for vms, while lxc uses container names/hostnames. So there is some extra work required.

So in some spots you have lxc-attach -n 201 on proxmox and lxc-attach -n somecontainer on LXC.

**Comment by g-bougard** [comment]
Hi @yoshi314 

sorry I don't understand why you want to discuss about obsolete code here. If you still want to be stuck with fia, you should know glpi-agent can be installed at the same time without problem. So you can apply the patch on the right place and not on fia and test it with more up-to-date code. Your issue with fia may still has been addressed, so I won't discuss it.

Then I still don't know if I should include the patch in next release...

**Comment by yoshi314** [comment]
well, it was just my use case since i actively use glpi+fa and we're switching over to glpi-agent.  But this will take a while.

And since codebase was relatively similar, i just worked with what i had at the time. 



I will try installing glpi-agent locally to reproduce the patch, just reported what has to be amended in the current one. The patch will be very similar, i just noted what has to be fixed in this one.

**Comment by yoshi314** [comment]
I'll try patching up glpi-agent today and send over the patch.

**Comment by yoshi314** [comment]
Ok, i have it working. I'll post a patch against latest debian release Lxc.pm file.

**Comment by yoshi314** [comment]
```
--- /root/Lxc.pm.org    2023-11-13 10:02:30.441421771 +0100
+++ /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Virtualization/Lxc.pm   2023-11-13 10:28:58.088395322 +0100
@@ -14,7 +14,7 @@
 use GLPI::Agent::Tools::Virtualization;
 
 sub isEnabled {
-    return canRun('lxc-ls');
+    return canRun('lxc-ls') || canRun('pct');
 }
 
 sub doInventory {
@@ -124,8 +125,12 @@
 sub  _getVirtualMachines {
     my (%params) = @_;
 
+    # Check if we require to list containers using proxmox pct command
+    my $use_pct = canRun('pct');
+    my $pct_name = 0;
+
     my @lines = getAllLines(
-        command => 'lxc-ls -1',
+        command => $use_pct ? 'pct list' : 'lxc-ls -1',
         %params
     );
     return unless @lines;
@@ -136,6 +141,7 @@
         %params
     );
 
+    my $vmid;
     my $lxcpath = getFirstLine(
         command => "lxc-config lxc.lxcpath",
         %params
@@ -147,12 +153,35 @@
     my @machines;
 
     foreach my $name (@lines) {
+        # Support pct when running with proxmox
+        if ($use_pct) {
+            if ($name =~ /^(VMID\s.*\s)Name.*$/) {
+                $pct_name = length($1);
+                next;
+            } elsif ($pct_name) {
+               if ( $name =~ m/(\d+)/ ) { 
+                       $vmid = $1;
+               }
+                $name = substr($name, $pct_name);
+            } else {
+                next;
+            }
+        }
+
+       my $ctid;
+
+       if ($use_pct) { # proxmox
+               $ctid = $vmid;
+       } else {   #lxc
+               $ctid = $name;
+       }
+
         # lxc-ls -1 shows one entry by line
         $name =~ s/\s+$//;         # trim trailing whitespace
         next unless length($name); # skip if empty as name can contain space
 
         my $container = _getVirtualMachine(
-            name    => $name,
+            name    => $use_pct ? $vmid : $name,
             version => $version,
             config  => "$lxcpath/$name/config",
             logger  => $params{logger}
@@ -168,18 +197,18 @@
         my ($machineid, $hostname);
         if ( $container->{STATUS} && $container->{STATUS} eq STATUS_RUNNING ) {
             $machineid = getFirstLine(
-                command => "lxc-attach -n '$name' -- /bin/cat /etc/machine-id",
+                command => "lxc-attach -n '$ctid' -- /bin/cat /etc/machine-id",
                 logger => $params{logger}
             );
             $hostname = getFirstLine(
-                command => "lxc-attach -n '$name' -- /bin/cat /etc/hostname",
+                command => "lxc-attach -n '$ctid' -- /bin/cat /etc/hostname",
                 logger => $params{logger}
             );
         } else {
             # Try to directly access container filesystem for not powered container
             # Works for standard fs or overlay rootfs
             my $rootfs = getFirstMatch(
-                command => "/usr/bin/lxc-info -n '$name' -c $rootfs_conf",
+                command => "/usr/bin/lxc-info -n '$ctid' -c $rootfs_conf",
                 pattern => qr/^lxc\.rootfs.*\s*=\s*(.*)$/,
                 logger  => $params{logger}
             );
```


Sample output from Proxmox is slightly different to lxc-ls output. And containers are identified by numbers, 
for instance

```
VMID       Status     Lock         Name                
221        running                 mclautotests        
226        running                 mclterenpol         
229        running                 mckat1dev           
310        running                 mczam1dev-deb12     
511        running                 mcloudx18   
```



If you want some data for unit tests, let me know.

**Comment by yoshi314** [comment]
It could use some cleaning up, but it gets the job done.


```
(...)
    <VIRTUALMACHINES>
      <MAC>12:91:80:74:52:d1</MAC>
      <NAME>mclautotests</NAME>
      <STATUS>running</STATUS>
      <UUID>7e8a4542202e910a67c8a72c6e7608a60ad2536b</UUID>
      <VCPU>56</VCPU>
      <VMTYPE>lxc</VMTYPE>
    </VIRTUALMACHINES>
    <VIRTUALMACHINES>
      <MAC>de:d6:77:64:52:a5</MAC>
      <NAME>mclterenpol</NAME>
      <STATUS>running</STATUS>
      <UUID>548fe980a95f23b1dbff70820ea65a2f2b17a093</UUID>
      <VCPU>56</VCPU>
      <VMTYPE>lxc</VMTYPE>
    </VIRTUALMACHINES>
    <VIRTUALMACHINES>
      <MAC>96:b5:db:bd:77:50</MAC>
      <NAME>mckat1dev</NAME>
      <STATUS>running</STATUS>
      <UUID>cee654a6fc6a7ba8668bf377c91c512eb394849d</UUID>
      <VCPU>56</VCPU>
      <VMTYPE>lxc</VMTYPE>
    </VIRTUALMACHINES>
    <VIRTUALMACHINES>
      <MAC>16:4e:34:c7:54:22</MAC>
      <NAME>mczam1dev-deb12</NAME>
      <STATUS>running</STATUS>
      <UUID>83eb8e5e2acf147b373482dc065b38a08b063e08</UUID>
      <VCPU>56</VCPU>
      <VMTYPE>lxc</VMTYPE>
    </VIRTUALMACHINES>
    <VIRTUALMACHINES>
      <MAC>02:d0:4a:9a:1c:7b</MAC>
      <NAME>mcloudx18</NAME>
      <STATUS>running</STATUS>
      <UUID>946adc31f7988448e07445570e7ce36150da9140</UUID>
      <VCPU>56</VCPU>
      <VMTYPE>lxc</VMTYPE>
    </VIRTUALMACHINES>
(...)
```

**Comment by g-bougard** [comment]
Hi @yoshi314 

thank you very much for your work !

After checking your patch, I understood what I missed. So I prepared an updated one for current develop branch which does exactly the same thing than yours with few little fixes/updates:

Here are my essential changes:
* move `canRun('pct')` in caller to avoid call during tests
* `$vmid` is declared inside the loop to keep it uninitialized when starting a loop to avoid duplicating a vmid if the output is corrupted
* I simplified `$ctid` computing and use
* I passed `$ctid` in `%params` to `_getVirtualMachine()` to keep the found name from `pct` output while initializing the `$container` hash. This changes nothing for you as this function still had code to find the right name set in config by Proxmox. So this will avoid a problem if utsname is no more set by proxmox in the future.
* I also updated `_getVirtualMachine()` to support `$ctid` and `$lxcpath` to find lxc config file when looking for NAME, MAC, VCPU or MEMORY

It will be integrated in next nightly build.
